pkg/controller/agent/fleet.go

```diff
@@ -276,22 +308,33 @@ func reconcileEnrollmentToken(ctx context.Context, agent agentv1alpha1.Agent, cl
 		}
 	}
 
-CREATE:
-	key, err := api.createEnrollmentAPIKey(ctx, policyID)
+FindOrCreate:
+	// try to re-use an existing key
+	newkey := false
+	key, err := api.findEnrollmentAPIKey(ctx, policyID)
 	if err != nil {
-		return EnrollmentAPIKey{}, err
+		ulog.FromContext(ctx).Info("Could not find existing Fleet enrollment API keys, creating new one", "error", err.Error())
+		key, err = api.createEnrollmentAPIKey(ctx, policyID)
+		if err != nil {
+			return EnrollmentAPIKey{}, err
+		}
+		newkey = true
 	}
+
 	// this potentially creates conflicts we could introduce reconciler state similar to the ES controller and handle it
 	// on the top level but we would then potentially create redundant enrollment tokens in the Fleet API
 	if agent.Annotations == nil {
 		agent.Annotations = map[string]string{}
 	}
 	agent.Annotations[FleetTokenAnnotation] = key.ID
 	err = client.Update(ctx, &agent)
-	if err != nil {
+	if err != nil && newkey {
 		// we have failed to store the token id in an annotation let's try to remove the token again
 		return EnrollmentAPIKey{}, k8serrors.NewAggregate([]error{err, api.deleteEnrollmentAPIKey(ctx, key.ID)})
```

Comment:
Is it still required since we attempt to reuse existing keys?

Reply:
You mean deleting the key? Hm you are right, we would try to reuse it next time round. I can remove that.